File: express-backend/src/services/provider/TraktMovieSearchProvider.ts
import axios from 'axios';
import { config } from '../../config';
import { prisma } from '../../config/db';
import { MovieSearchProvider } from '../provider/MovieSearchProvider';
import {
  ExternalMovieSearchResult,
  ProviderSearchResult,
  MovieRecordSuggestion,
} from '../../dto/external-search';
import { RecordStatus } from '../../enums/RecordStatus';

// Trakt 影视搜索 Provider，与 Java 端 TraktMovieSearchProvider 完全对齐
export class TraktMovieSearchProvider implements MovieSearchProvider {
  private static readonly PAGE_SIZE = 20;

  id(): string {
    return 'trakt';
  }

  async search(query: string, page: number): Promise<ProviderSearchResult<ExternalMovieSearchResult>> {
    const result: ProviderSearchResult<ExternalMovieSearchResult> = {
      provider: this.id(),
      enabled: true,
      message: '',
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };

    if (!config.trakt.clientId) {
      result.enabled = false;
      result.message = 'Trakt 未启用';
      return result;
    }

    if (!query) throw new Error('query must not be blank');
    const normalizedPage = Math.max(page, 1);

    const response = await axios.get(`${config.trakt.baseUrl}/search/movie`, {
      params: { query, page: normalizedPage, limit: TraktMovieSearchProvider.PAGE_SIZE },
      headers: {
        'trakt-api-key': config.trakt.clientId,
        'trakt-api-version': '2',
      },
    });

    const items = response.data ?? [];

    // 批量查本地数据库
    const traktIds: string[] = [];
    const tmdbIds: number[] = [];
    const imdbIds: string[] = [];
    for (const item of items) {
      if (item.movie?.ids?.trakt) traktIds.push(String(item.movie.ids.trakt));
      if (item.movie?.ids?.tmdb) tmdbIds.push(item.movie.ids.tmdb);
      if (item.movie?.ids?.imdb) imdbIds.push(item.movie.ids.imdb);
    }

    const byTraktId = traktIds.length > 0
      ? await this.findByField('traktId', traktIds) : new Map<string, any>();
    const byTmdbId = tmdbIds.length > 0
      ? await this.findByFieldBigInt('tmdbId', tmdbIds) : new Map<number, any>();
    const byImdbId = imdbIds.length > 0
      ? await this.findByField('imdbId', imdbIds) : new Map<string, any>();

    const results: ExternalMovieSearchResult[] = [];
    for (const item of items) {
      if (!item.movie || !item.movie.ids) continue;
      const ids = item.movie.ids;
      const mapped: ExternalMovieSearchResult = {
        provider: this.id(),
        title: item.movie.title ?? '',
        releaseDate: item.movie.year != null ? String(item.movie.year) : null,
        traktId: ids.trakt != null ? String(ids.trakt) : null,
        imdbId: ids.imdb ?? null,
        tmdbId: ids.tmdb ?? null,
        doubanId: null,
        posterUrl: null,
        overview: null,
        alreadyAdded: false,
        existingRecordId: null,
        suggestedRecord: null,
      };

      // 查本地是否已添加
      let existing: any = null;
      if (mapped.traktId) existing = byTraktId.get(mapped.traktId);
      if (!existing && mapped.tmdbId) existing = byTmdbId.get(mapped.tmdbId);
      if (!existing && mapped.imdbId) existing = byImdbId.get(mapped.imdbId);
      mapped.alreadyAdded = existing !== null && existing !== undefined;
      mapped.existingRecordId = existing?.id != null ? Number(existing.id) : null;
      mapped.suggestedRecord = this.buildSuggestion(mapped);
      results.push(mapped);
    }

    result.page = normalizedPage;
    result.totalPages = 0;
    result.totalResults = results.length;
    result.results = results;
    return result;
  }

  private async findByField(field: string, values: string[]): Promise<Map<string, any>> {
    const where: any = { [field]: { in: values } };
    const movies = await prisma.movie.findMany({ where });
    return new Map(movies.map((m: any) => [m[field], m]));
  }

  private async findByFieldBigInt(field: string, values: number[]): Promise<Map<any, any>> {
    const where: any = { [field]: { in: values } };
    const movies = await prisma.movie.findMany({ where });
    return new Map(movies.map((m: any) => [m[field], m]));
  }

  private buildSuggestion(mapped: ExternalMovieSearchResult): MovieRecordSuggestion {
    return {
      tmdbId: mapped.tmdbId,
      imdbId: mapped.imdbId,
      doubanId: mapped.doubanId,
      traktId: mapped.traktId,
      title: mapped.title,
      posterUrl: mapped.posterUrl,
      status: RecordStatus.WANT,
      rating: null,
      shortReview: '',
    };
  }
}

File: legacy/java-backend/src/main/java/com/pixelreel/service/provider/TraktMovieSearchProvider.java
package com.pixelreel.service.provider;

import com.pixelreel.config.TraktProperties;
import com.pixelreel.dto.external.ExternalMovieSearchResult;
import com.pixelreel.dto.external.MovieRecordSuggestion;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.dto.trakt.TraktSearchItem;
import com.pixelreel.entity.Movie;
import com.pixelreel.enums.RecordStatus;
import com.pixelreel.service.MovieService;
import com.pixelreel.service.search.MovieSearchProvider;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class TraktMovieSearchProvider implements MovieSearchProvider {
  private static final int DEFAULT_PAGE_SIZE = 20;

  private final MovieService movieService;

  @Qualifier("traktRestClient")
  private final RestClient traktRestClient;

  private final TraktProperties traktProperties;

  @Override
  public String id() {
    return "trakt";
  }

  @Override
  public ProviderSearchResult<ExternalMovieSearchResult> search(String query, int page) {
    ProviderSearchResult<ExternalMovieSearchResult> result = new ProviderSearchResult<>();
    result.setProvider(id());

    if (!traktProperties.enabled()) {
      result.setEnabled(false);
      result.setMessage("Trakt 未启用");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(query)) {
      throw new IllegalArgumentException("query must not be blank");
    }

    List<TraktSearchItem> items = traktRestClient
        .get()
        .uri(uriBuilder -> uriBuilder
            .path("/search/movie")
            .queryParam("query", query)
            .queryParam("page", Math.max(page, 1))
            .queryParam("limit", DEFAULT_PAGE_SIZE)
            .build())
        .header("trakt-api-key", traktProperties.clientId())
        .header("trakt-api-version", String.valueOf(Optional.ofNullable(traktProperties.apiVersion()).orElse(2)))
        .retrieve()
        .body(new ParameterizedTypeReference<List<TraktSearchItem>>() {});

    List<TraktSearchItem> safeItems = items == null ? Collections.emptyList() : items;

    Map<String, Movie> byTraktId = findExistingByTraktId(safeItems);
    Map<Long, Movie> byTmdbId = findExistingByTmdbId(safeItems);
    Map<String, Movie> byImdbId = findExistingByImdbId(safeItems);

    List<ExternalMovieSearchResult> results = new ArrayList<>();
    for (TraktSearchItem item : safeItems) {
      if (item.getMovie() == null || item.getMovie().getIds() == null) {
        continue;
      }

      ExternalMovieSearchResult mapped = new ExternalMovieSearchResult();
      mapped.setProvider(id());
      mapped.setTitle(item.getMovie().getTitle());
      mapped.setReleaseDate(item.getMovie().getYear() == null ? null : String.valueOf(item.getMovie().getYear()));
      mapped.setTraktId(item.getMovie().getIds().getTrakt() == null ? null : String.valueOf(item.getMovie().getIds().getTrakt()));
      mapped.setImdbId(item.getMovie().getIds().getImdb());
      mapped.setTmdbId(item.getMovie().getIds().getTmdb());

      Movie existing = findExisting(mapped, byTraktId, byTmdbId, byImdbId);
      mapped.setAlreadyAdded(existing != null);
      mapped.setExistingRecordId(existing == null ? null : existing.getId());
      mapped.setSuggestedRecord(buildSuggestion(mapped));
      results.add(mapped);
    }

    result.setEnabled(true);
    result.setPage(Math.max(page, 1));
    result.setTotalPages(0);
    result.setTotalResults(results.size());
    result.setResults(results);
    return result;
  }

  private Movie findExisting(
      ExternalMovieSearchResult mapped,
      Map<String, Movie> byTraktId,
      Map<Long, Movie> byTmdbId,
      Map<String, Movie> byImdbId
  ) {
    if (StringUtils.hasText(mapped.getTraktId())) {
      Movie existing = byTraktId.get(mapped.getTraktId());
      if (existing != null) {
        return existing;
      }
    }
    if (mapped.getTmdbId() != null) {
      Movie existing = byTmdbId.get(mapped.getTmdbId());
      if (existing != null) {
        return existing;
      }
    }
    if (StringUtils.hasText(mapped.getImdbId())) {
      return byImdbId.get(mapped.getImdbId());
    }
    return null;
  }

  private Map<String, Movie> findExistingByTraktId(List<TraktSearchItem> items) {
    List<Long> ids = items.stream()
        .map(item -> item.getMovie() == null || item.getMovie().getIds() == null
            ? null
            : item.getMovie().getIds().getTrakt())
        .filter(value -> value != null)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return movieService.lambdaQuery()
        .in(Movie::getTraktId, ids.stream().map(String::valueOf).toList())
        .list()
        .stream()
        .collect(Collectors.toMap(Movie::getTraktId, Function.identity(), (a, b) -> a));
  }

  private Map<Long, Movie> findExistingByTmdbId(List<TraktSearchItem> items) {
    List<Long> ids = items.stream()
        .map(item -> item.getMovie() == null || item.getMovie().getIds() == null
            ? null
            : item.getMovie().getIds().getTmdb())
        .filter(value -> value != null)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return movieService.lambdaQuery()
        .in(Movie::getTmdbId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Movie::getTmdbId, Function.identity(), (a, b) -> a));
  }

  private Map<String, Movie> findExistingByImdbId(List<TraktSearchItem> items) {
    List<String> ids = items.stream()
        .map(item -> item.getMovie() == null || item.getMovie().getIds() == null
            ? null
            : item.getMovie().getIds().getImdb())
        .filter(StringUtils::hasText)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return movieService.lambdaQuery()
        .in(Movie::getImdbId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Movie::getImdbId, Function.identity(), (a, b) -> a));
  }

  private MovieRecordSuggestion buildSuggestion(ExternalMovieSearchResult mapped) {
    MovieRecordSuggestion suggestion = new MovieRecordSuggestion();
    suggestion.setTmdbId(mapped.getTmdbId());
    suggestion.setImdbId(mapped.getImdbId());
    suggestion.setTraktId(mapped.getTraktId());
    suggestion.setTitle(mapped.getTitle());
    suggestion.setPosterUrl(mapped.getPosterUrl());
    suggestion.setStatus(RecordStatus.WANT);
    suggestion.setRating(null);
    suggestion.setShortReview("");
    return suggestion;
  }

  
}


